File: 3-HARD/Python/1-travelling-salesman-bitmasking/proposed-solution/main.py
# Challenge: Implement a solution for the travelling salesman problem using dynamic programming with bitmasking.
# This challenge requires optimizing the search in an exponential state space.
# Use object-oriented programming and follow the DRY principle.

from tsp import TravellingSalesman

def main():
    graph = [
        [0, 10, 15, 20],
        [10, 0, 35, 25],
        [15, 35, 0, 30],
        [20, 25, 30, 0]
    ]
    tsp_solver = TravellingSalesman(graph)
    min_cost = tsp_solver.tsp(0, 1 << 0)
    print(f"Minimum travelling salesman cost: {min_cost}")

if __name__ == "__main__":
    main()

File: 3-HARD/JavaScript/1-travelling-salesman-bitmasking/proposed-solution/main.js
/*
Challenge:  
Given a set of cities and the distances between every pair of cities, find the shortest possible route that visits each city exactly once and returns to the origin city.  
This challenge requires implementing the Travelling Salesman Problem (TSP) using bitmasking and dynamic programming to optimize the solution.
*/

import TravellingSalesman from "./TravellingSalesman.js";

function main() {
  const distanceMatrix = [
    [0, 10, 15, 20],
    [10, 0, 35, 25],
    [15, 35, 0, 30],
    [20, 25, 30, 0],
  ];

  const tspSolver = new TravellingSalesman(distanceMatrix);
  const minCost = tspSolver.tsp();
  console.log("Minimum cost to visit all cities:", minCost);
}

main();
